editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -1,284 +1,569 @@
-      newParsed.topLevelElements,
-      workingComponents,
-    )
-
-    const fixedHighlightBounds = mapArrayToDictionary(
-      Object.values(newParsed.highlightBounds),
-      (highlightBound) => {
-        const uidFix = newToOldUidMapping[highlightBound.uid]
-        if (uidFix != null) {
-          return uidFix.oldUID
-        } else {
-          return highlightBound.uid
-        }
-      },
-      (highlightBound) => {
-        const uidFix = newToOldUidMapping[highlightBound.uid]
-        if (uidFix != null) {
-          return {
-            ...highlightBound,
-            uid: uidFix.oldUID,
-          }
-        } else {
-          return highlightBound
-        }
-      },
-    )
 
-    return {
-      ...newParsed,
-      topLevelElements: fixedTopLevelElements,
-      highlightBounds: fixedHighlightBounds,
-    }
-  } else {
-    // we found no uid mismatch or we found too many mismatched UIDs and so we bail out
-    return newParsed
+  // This needs to be corrected as things may have moved around.
+  const fixedHighlightBounds: HighlightBoundsForUids = {
+    ...newParsed.highlightBounds,
+  }
+  for (const { oldUID, newUID } of fixUIDsState.mappings) {
+    const bounds = fixedHighlightBounds[oldUID]
+    delete fixedHighlightBounds[oldUID]
+    fixedHighlightBounds[newUID] = bounds
+  }
+
+  // Return the result.
+  return {
+    ...newParsed,
+    topLevelElements: fixedTopLevelElements,
+    combinedTopLevelArbitraryBlock: fixedCombinedTopLevelArbitraryBlock,
+    highlightBounds: fixedHighlightBounds,
   }
 }
 
-type OnElement = (
+function updateUID<T>(
+  uidOptic: Optic<T, string>,
   oldUID: string,
-  newUID: string,
-  oldPathToRestore: StaticElementPathPart,
-  newElementPath: StaticElementPathPart,
-) => void
-
-function zipTopLevelElements(
-  firstTopLevelElements: Array<TopLevelElement>,
-  secondTopLevelElements: Array<TopLevelElement>,
-  onElement: OnElement,
-): boolean {
-  const firstComponents = getComponentsFromTopLevelElements(firstTopLevelElements)
-  const secondComponents = getComponentsFromTopLevelElements(secondTopLevelElements)
-
-  const lengthsMatchAtThisLevel = firstComponents.length === secondComponents.length
-  let numberOfElementsMatch = lengthsMatchAtThisLevel
-
-  firstComponents.forEach((firstComponent, index) => {
-    if (secondComponents.length > index) {
-      const secondComponent = secondComponents[index]
-
-      const lengthsOfChildrenMatch = walkElementChildren(
-        EP.emptyElementPathPart,
-        [firstComponent.rootElement],
-        [secondComponent.rootElement],
-        onElement,
-      )
-      numberOfElementsMatch = numberOfElementsMatch && lengthsOfChildrenMatch
+  fixUIDsState: FixUIDsState,
+  baseValue: T,
+): T {
+  const newUID = unsafeGet(uidOptic, baseValue)
+  let uidToUse: string = oldUID
+  if (oldUID === newUID) {
+    // Old one is the same as the new one, so everything is great.
+    uidToUse = newUID
```

Comment:
this case can be removed, as it's already covered by the assignment on line `114`

Reply:
The code on line 114 sets it to `oldUID`...

Reply:
but they are checked to be the same, right?

Reply:
I've removed the assignment from line 114, so that it's mandatory to assign it in the various conditionals, in 6a3d04e93.